Disable -Ydelambdafy:method for specialized FunctionN

retronym opened the pull request:
The Delambdafy phase generates its `FunctionN` subclasses after
the specialization phase. As such, `((x: Int) => x).apply(42)` incurs
boxing.

This commit falls back to the `-Ydelambdafy:inline` in this case.

This is done by running the specialization type map over the
type of the function, and seeing if anything changes. To make this
work robustly, we first need to ensure that the specialization info
transformer has processed all the function types.

This is not a fundamental limitation; we could in principle generate
the specialized code.

A followup change will use `-Ydelambdafy:method` as the basis for
invokedymnamic lambdas. As part of that stream of
work, we will synthesize specialization-aware lambdas, and remove
the fallback to `-Ydelambdafy:inline`.

Review by @gkossakowski

I have updated some tests that indend to test the delambdafy transform
to avoid use of specialized function types.

retronym committed:
Disable -Ydelambdafy:method for specialized FunctionN

The Delambdafy phase generates its `FunctionN` subclasses after
the specialization phase. As such, `((x: Int) => x).apply(42)` incurs
boxing.

This commit falls back to the `-Ydelambdafy:inline` in this case.

This is done by running the specialization type map over the
type of the function, and seeing if anything changes. To make this
work robustly, we first need to ensure that the specialization info
transformer has processed all the function types.

This is not a fundamental limitation; we could in principle generate
the specialized code.

A followup change will use `-Ydelambdafy:method` as the basis for
invokedymnamic lambdas. As part of that stream of
work, we will synthesize specialization-aware lambdas, and remove
the fallback to `-Ydelambdafy:inline`.

I have updated some tests that intend to test the delambdafy transform
to avoid use of specialized function types.

gkossakowski commented:
It looks like this failure is legit:

```
java.lang.AssertionError: assertion failed: [[syntax trees at end of                delambdafy]] // newSource1.scala
package o {
  package a {
    class C extends Object {
      def hihi(): List = immutable.this.List.apply(scala.this.Predef.wrapIntArray(Array[Int]{1, 2})).map({
  (new <$anon: Function1>(C.this): Function1)
}, immutable.this.List.canBuildFrom()).$asInstanceOf[List]();
      def <init>(): o.a.C = {
        C.super.<init>();
        ()
      }
    };
    @SerialVersionUID(value = 0) final <synthetic> class anonfun$hihi$1 extends scala.runtime.AbstractFunction1$mcII$sp with Serializable {
      final def apply(x$1: Int): Int = anonfun$hihi$1.this.apply$mcII$sp(x$1);
      <specialized> def apply$mcII$sp(x$1: Int): Int = x$1.*(2);
      final <bridge> <artifact> def apply(v1: Object): Object = scala.Int.box(anonfun$hihi$1.this.apply(scala.Int.unbox(v1)));
      def <init>($outer: o.a.C): <$anon: Function1> = {
        anonfun$hihi$1.super.<init>();
        ()
      }
    }
  };
  package a {
    object `package` extends Object {
      def f(): Int = 1;
      def <init>(): o.a.package.type = {
        `package`.super.<init>();
        ()
      }
    }
  }
}


    at Test$.show(t9097.scala:32)
```

retronym commented:
That was a newly added test in the category of test that "I updated some tests that intend to test the delambdafy transform to avoid use of specialized function types."

I've just changed it to use a non-specialized function so that delambdafy is still used.

retronym commented:
Just tweaked another of a test that intends to exercise `-Ydelambdafy:method`, but was using a `(Int => Int)` function whose encoding differs after this PR.

gkossakowski commented:
/rebuild

I'm not sure why those tests are failing.

gkossakowski commented on src/compiler/scala/tools/nsc/transform/UnCurry.scala:
```diff
@@ -70,6 +70,14 @@ abstract class UnCurry extends InfoTransform
     private val newMembers        = mutable.Map[Symbol, mutable.Buffer[Tree]]()
 
+    private lazy val forceSpecializationInfoTransformOfFunctionN: Unit = {
+      if (currentRun.specializePhase != NoPhase) {
```
Why is this needed?

retronym replied:
I don't recall. Maybe it is for custom pipelines like ScalaJS.

gkossakowski replied:
If it's not clear why it's needed, should we remove it? This kind of code is hard to remove later on because everybody fears some hard to predict breakage.

retronym commented:
@gkossakowski By removing that condition, I found that a test case using `-Ystop-after` uncurry failed. That must have been what prompted by to add this code in the first place. I have reinstated it with a supporting comment.

retronym commented:
@adriaanm / @lrytz could I ask for a review of this one?

adriaanm commented:
LGTM -- Merging to unlock next steps. There's a small typo btw. Left as exercise for the submitter :)

adriaanm commented:
I was just about to hit the merge button, @retronym -- meant to do sooner, sorry.

adriaanm commented:
PS: canUseDelam_b_dafyMethod
